packages/fs-detectors/src/detect-framework.ts

```diff
@@ -90,3 +150,83 @@ export async function detectFramework({
   );
   return result.find(res => res !== null) ?? null;
 }
+
+type VersionedFramework = Framework & {
+  detectedVersion?: string;
+};
+
+// Note: Does not currently support a `frameworkList` of monorepo managers
+export async function detectFrameworkRecord({
+  fs,
+  frameworkList,
+}: DetectFrameworkRecordOptions): Promise<VersionedFramework | null> {
+  const result = await Promise.all(
+    frameworkList.map(async frameworkMatch => {
+      const matchResult = await matches(fs, frameworkMatch);
+      if (matchResult) {
+        return {
+          ...frameworkMatch,
+          detectedVersion: matchResult?.detectedVersion,
+        };
+      }
+      return null;
+    })
+  );
+  const frameworkRecord = result.find(res => res !== null) ?? null;
+
+  if (!frameworkRecord) {
+    return null;
+  }
+
+  if (frameworkRecord.detectedVersion) {
+    return frameworkRecord;
+  }
+
+  const detectedVersion = await detectFrameworkVersion(frameworkRecord);
+  if (!detectedVersion) {
+    return frameworkRecord;
+  }
+
+  return {
+    ...frameworkRecord,
+    detectedVersion,
+  };
+}
+
+export async function detectFrameworkVersion(
+  frameworkRecord: Framework
+): Promise<string | undefined> {
+  const allDetectors = [
+    ...(frameworkRecord.detectors?.every || []),
+    ...(frameworkRecord.detectors?.some || []),
+  ];
+  const firstMatchPackage = allDetectors.find(d => d.matchPackage);
+
+  if (!firstMatchPackage?.matchPackage) {
+    return;
+  }
+
+  const version = lookupInstalledVersion(firstMatchPackage.matchPackage);
+  if (version) {
+    return version;
+  }
+
+  return;
+}
+
+function lookupInstalledVersion(packageName: string): string | undefined {
+  try {
+    // we can't use "require(`${packageName}/package.json`)" here because
+    // ncc bundling will make it look in the wrong place
+    const script = `eval('require')('${packageName}/package.json').version`;
```

Comment:
`eval` worries me. Lets be extra diligent this is not introducing some kind of code injection vulnerability

Reply:
`eval()` is necessary to get the global Node.js `require()` and not the webpack version. Webpack's `require()` can only return modules that are bundled. In this case, the code needs `package.json` which is not included in the bundle, thus Node.js' `require()` will easily resolve it.

Reply:
Yeah that makes sense. Just calling it out here that we need to be extra careful this doesn't introduce a vulnerability. I'm not an expert in that field but `eval` always sets off the alarm bells for me.

Reply:
I believe `eval('required')` is necessary because of some ncc processing.
https://github.com/vercel/vercel/pull/9009#discussion_r1051235678

Reply:
This is actually a bit more interesting. I thought the `spawnSync()` was a hack to get around the bundler and thought it could be removed in favor of the `eval('require')`, but it turns out the `eval('require')('<package>/package.json')` trick doesn't work! You have to spawn a subprocess.

I did some testing and both the bundled output and the Node.js REPL error executing `require('gatsby/package.json')`. However when I run `node -p "require('gatsby/package.json').version"` as you have it, it works! I have no idea why. They shouldn't be any different, right?

Additionally, it doesn't appear that the `eval()` is necessary after all. For the sake of clarity, I vote to remove the `eval()`.